src/com/zulip/android/HTTPRequest.java

```diff
@@ -1,137 +1,150 @@
-            String url = app.getServerURI() + path;
-            nameValuePairs.add(new BasicNameValuePair("client", "Android"));
-
-            if (method.equals("POST")) {
-                request = new HttpPost(url);
-            } else if (method.equals("PUT")) {
-                request = new HttpPut(url);
-            } else if (method.equals("DELETE")) {
-                request = new HttpDeleteWithReq(url);
-            } else if (method.equals("GET")) {
-                request = new HttpGet(url + "?"
-                        + URLEncodedUtils.format(nameValuePairs, "utf-8"));
+            URL url = new URL(app.getServerURI() + path);
+            this.setProperty("client", "Android");
+            String authstr = this.app.getEmail() + ":" + this.app.getApiKey();
+
+            if (!method.equals("POST") && !method.equals("DELETE") &&
+                    !method.equals("PUT") && !method.equals("GET")){
+                throw new IOException("Wrong HTTP method.");
+            }
+
+            if (method.equals("GET")) {
+                url = new URL(app.getServerURI() + path + "?" + getPostDataString(this.postDataParams));
             }
 
+            conn = (HttpsURLConnection) url.openConnection();
+            conn.setRequestMethod(method);
+            conn.setRequestProperty("Authorization", "Basic "
+                    + Base64.encodeToString(authstr.getBytes(), Base64.NO_WRAP));
+            conn.setRequestProperty("User-Agent", app.getUserAgent());
+            conn.setUseCaches(false);
+
             if (!method.equals("GET")) {
-                ((HttpEntityEnclosingRequestBase) request)
-                        .setEntity(new UrlEncodedFormEntity(nameValuePairs,
-                                "UTF-8"));
+                conn.setDoInput(true);
+                conn.setDoOutput(true);
+                DataOutputStream request = new DataOutputStream(conn.getOutputStream());
+                String parameters = getPostDataString(this.postDataParams);
+                Log.i("HTTP.parameters", parameters);
+                request.writeBytes(parameters);
+                request.flush();
+                request.close();
             }
 
-            Log.i("HTTP.request", request.getMethod() + " " + request.getURI());
+            Log.i("HTTP.request", conn.getRequestMethod() + " " + conn.getURL());
 
-            String authstr = this.app.getEmail() + ":" + this.app.getApiKey();
-            request.setHeader(
-                    "Authorization",
-                    "Basic "
-                            + Base64.encodeToString(authstr.getBytes(),
-                                    Base64.NO_WRAP));
             // timeout after 60 seconds in ms
-            HttpConnectionParams.setSoTimeout(request.getParams(), 60 * 1000);
+            conn.setConnectTimeout(60 * 1000);
+            conn.connect();
+
+            InputStream responseStream;
+
+            if(conn.getResponseCode() >= HttpStatus.SC_BAD_REQUEST)
```

Comment:
We should move off `HttpStatus` if possible, since leaving it in requires us to continue using deprecated libs.